### PR TITLE
Implementation of the `pathreparam` integrator with OptiX 7

### PR DESCRIPTION
**Pull request by Speierers**
Continuation of PR #44.

### CODE REVIEW

**Comment by abhinavvs** [comment]
Does this mean that "pathreparam" plug-in still is not fully released yet (not even on OptiX 6)? 

Any rough idea on when this plug-in will be available (and fully integrated with Mitsuba2)? I am really looking forward to this plug-in as I need this feature for my research project. 

Thanks!

**Comment by Speierers** [comment]
The previous implementation based on "Optix 6" wasn't release, although it is publicly available here:
https://github.com/loubetg/mitsuba2

**Comment by realWDC** [comment]
Great works!

Quick question:
How to auto-diff w.r.t. shape parameters, for example `m_radius` in `src/shapes/sphere.cpp`?

I have seen following comments in this PR; Any hints on how to proceed?
```cpp
 // TODO: make si differentiable w.r.t. shape parameters if necessary
```
Thanks!

**Comment by Speierers** [comment]
> Great works!
> 
> Quick question:
> How to auto-diff w.r.t. shape parameters, for example `m_radius` in `src/shapes/sphere.cpp`?
> 
> I have seen following comments in this PR; Any hints on how to proceed?
> 
> ```c++
>  // TODO: make si differentiable w.r.t. shape parameters if necessary
> ```
> 
> Thanks!

This is not supported at the moment. For this to work, `Shape::m_to_world` would need to become a differentiable type (e.g. `Transform<Point<CUDAArray<float>, 4>>` which involves a lot of changes accross the different shape plugins.
This will definitively be part of a future PR, but I won't be able to give any ETA at the moment.

File: src/shapes/disk.cpp [hunk 1]
```diff
@@ -208,16 +220,24 @@ class Disk final : public Shape<Float, Spectrum> {
...
+        si.p[active]            = ray_(si.t);
+        masked(si.time, active) = ray_.time;
+
+        return si;
+    }
 
-        si_out[active] = si;
+    std::pair<Point3f, Normal3f> differentiable_position(const SurfaceInteraction3f &si,
+                                                         Mask /*active*/) const override {
+        auto [c, s] = sincos(si.uv.y() * math::TwoPi<Float>);
```
Comment (realWDC) [inline src/shapes/disk.cpp:237]: Should it be `[s, c]` instead of `[c, s]`?
```cpp
auto [c, s] = sincos(si.uv.y() * math::TwoPi<Float>); // wrong?
auto [s, c] = sincos(si.uv.y() * math::TwoPi<Float>); // correct?
```

File: src/librender/mesh.cpp [hunk 2]
```diff
@@ -396,13 +399,40 @@ Mesh<Float, Spectrum>::barycentric_coordinates(const SurfaceInteraction3f &si,
...
+MTS_VARIANT typename Mesh<Float, Spectrum>::SurfaceInteraction3f
+Mesh<Float, Spectrum>::fill_surface_interaction(const Ray3f &ray,
+                                                const Float *cache,
+                                                const UInt32 &cache_indices,
+                                                SurfaceInteraction3f si,
+                                                Mask active) const {
+    MTS_MASK_ARGUMENT(active);
+
+    // Check whether the SurfaceInteraction need to be differentiable w.t.r. m_vertex_positions_buf
+    bool differentiable_pos = is_diff_array_v<Float> && requires_gradient(m_vertex_positions_buf);
```
Comment (realWDC) [inline src/librender/mesh.cpp:411]: To enable `scalar_*` variants compilation, how about changing from
```cpp
bool differentiable_pos = is_diff_array_v<Float> && requires_gradient(m_vertex_positions_buf);
```
to
```cpp
bool differentiable_pos = false;
if constexpr (is_diff_array_v<Float>) {
    differentiable_pos = requires_gradient(m_vertex_positions_buf);
}
```

**Review by wjakob — COMMENTED** [review]
This is a a big PR, which makes this very difficult to review. Can we please split it up into smaller parts with the key parts?

Specifically:

- differentiable intersections
- other things needed by this plugin (smootharea, etc.)
- the plugin itself

Since these depend on each other, we could do them one after the other instead of having 3 parallel PRs that would be difficult to keep in sync.

For the differentiable intersections, it will be essential to have testcases to make sure that the differentiation is working as expected. (Imagine how difficult it would be to investigate a poorly converging gradient descent, only to find that derivatives are not correctly propagated between mesh positions and derived shading normals.)

For the first part, it would also be good to discuss some of the design decisions that were made. For instance, how `fill_intersection` gets an intersection record as input and fills it with further information? I understand why this was done due to difficulties in modifying parameters via references in the GPU backends. But there are also downsides to this approach: the masking that this is now needed in all implementations of this method (a source of errors when a mask is forgotten somewhere?), along with the general costs of passing around this data structure (it’s the largest one we have in Mitsuba2). What I would suggest is that we develop an interface where the input SurfaceInteraction3f is not needed.

**Comment by priyamvad** [comment]
Hello, any movement on this PR? I'd love to test out the ability to derive a displacement map to match the reference render, something it sounds like this feature should help with?

**Comment by Speierers** [comment]
> Hello, any movement on this PR? I'd love to test out the ability to derive a displacement map to match the reference render, something it sounds like this feature should help with?

Hi @priyamvad,
We are currently busy working on a major refactoring of the Mitsuba 2 + Enoki codebase. 

You should be able to use this branch as it is for you experiments. Have a look at the ` heightfield` optimization tutorial to get started.

**Comment by RiverIntheSky** [comment]
> > Great works!
> > Quick question:
> > How to auto-diff w.r.t. shape parameters, for example `m_radius` in `src/shapes/sphere.cpp`?
> > I have seen following comments in this PR; Any hints on how to proceed?
> > ```c++
> >  // TODO: make si differentiable w.r.t. shape parameters if necessary
> > ```
> > 
> > 
> > Thanks!
> 
> This is not supported at the moment. For this to work, `Shape::m_to_world` would need to become a differentiable type (e.g. `Transform<Point<CUDAArray<float>, 4>>` which involves a lot of changes accross the different shape plugins.
> This will definitively be part of a future PR, but I won't be able to give any ETA at the moment.

Excuse me, is there any update on differentiable to_world? Is this pathreparam branch necessary for this goal compared to master branch? If I'm going to proceed in this direction myself, where should I make changes? Currently I can't pass Transform4f to optix::Transform4f

**Comment by Speierers** [comment]
> > > Great works!
> > > Quick question:
> > > How to auto-diff w.r.t. shape parameters, for example `m_radius` in `src/shapes/sphere.cpp`?
> > > I have seen following comments in this PR; Any hints on how to proceed?
> > > ```c++
> > >  // TODO: make si differentiable w.r.t. shape parameters if necessary
> > > ```
> > > 
> > > 
> > > Thanks!
> > 
> > 
> > This is not supported at the moment. For this to work, `Shape::m_to_world` would need to become a differentiable type (e.g. `Transform<Point<CUDAArray<float>, 4>>` which involves a lot of changes accross the different shape plugins.
> > This will definitively be part of a future PR, but I won't be able to give any ETA at the moment.
> 
> Excuse me, is there any update on differentiable to_world? Is this pathreparam branch necessary for this goal compared to master branch? If I'm going to proceed in this direction myself, where should I make changes? Currently I can't pass Transform4f to optix::Transform4f

We are busy working on a major refactoring and haven't made any progress on this unfortunately.

Differentiable `to_world` will for sure introduce discontinuities in the rendering equation, therefore using `pathreparam` will be necessary to account for those.

You will need to make changes to the shape class and the shape plugins so that there `traverse()` and `parameters_changed()` method expose and handle changes in `m_to_world`.

To pass the transform to optix, you will need to evaluate it and reduce it to a scalar value (e.g. using `enoki::hsum` which won't do much as every entry of that matrix is a single value).

Good luck!

**Comment by RiverIntheSky** [comment]
> > > > Great works!
> > > > Quick question:
> > > > How to auto-diff w.r.t. shape parameters, for example `m_radius` in `src/shapes/sphere.cpp`?
> > > > I have seen following comments in this PR; Any hints on how to proceed?
> > > > ```c++
> > > >  // TODO: make si differentiable w.r.t. shape parameters if necessary
> > > > ```
> > > > 
> > > > 
> > > > Thanks!
> > > 
> > > 
> > > This is not supported at the moment. For this to work, `Shape::m_to_world` would need to become a differentiable type (e.g. `Transform<Point<CUDAArray<float>, 4>>` which involves a lot of changes accross the different shape plugins.
> > > This will definitively be part of a future PR, but I won't be able to give any ETA at the moment.
> > 
> > 
> > Excuse me, is there any update on differentiable to_world? Is this pathreparam branch necessary for this goal compared to master branch? If I'm going to proceed in this direction myself, where should I make changes? Currently I can't pass Transform4f to optix::Transform4f
> 
> We are busy working on a major refactoring and haven't made any progress on this unfortunately.
> 
> Differentiable `to_world` will for sure introduce discontinuities in the rendering equation, therefore using `pathreparam` will be necessary to account for those.
> 
> You will need to make changes to the shape class and the shape plugins so that there `traverse()` and `parameters_changed()` method expose and handle changes in `m_to_world`.
> 
> To pass the transform to optix, you will need to evaluate it and reduce it to a scalar value (e.g. using `enoki::hsum` which won't do much as every entry of that matrix is a single value).
> 
> Good luck!

I am not sure if I understand `enoki::hsum` correctly. As I read in [Enoki Reference](https://enoki.readthedocs.io/en/master/reference.html?highlight=hsum#_CPPv4I0EN5enoki4hsumE7value_tI5ArrayE5Array), `value_t<Array> hsum(Array value)`, so `hsum(Float m_radius)` should return `value_t<Float>`, which I tested is `float` (my compiler says `'value_t<enoki::DiffArray<CUDAArray<float>>>' (aka 'float')`. But as I tried to assign it to `float` it failed:
```
error: no viable conversion from 'enoki::DiffArray<CUDAArray<float>>' to 'float'
        float radius = hsum(m_radius);
              ^        ~~~~~~~~~~~~~~
```
Why is this? It looks like `hsum(Float m_radius)` still returns `Float`.

**Comment by Speierers** [comment]
Hi @RiverIntheSky ,

Could you please open a seperate issue for this? Could you try something like `hsum(detach(m_radius))` maybe?

**Comment by idiot-z** [comment]
Dear developers,
First of all thank you for your great tool and all the work done.
If I understand correctly differentiation of weights for blended BSDFs as well as mesh positions depend on this PR? Or was it already included as part of another refactoring mentioned above?
Do you have an idea if work on this PR will start again in the near future?
Thank you very much for your time.

**Comment by Speierers** [comment]
Hi @idiot-z ,

This PR will stay as it is for now. The refactored codebase will be very different, requiring a complete re-write of this plugin to fully leverage the various new features of the system. I can't say for now whether it will included or not "natively" in the upcoming release.

**Comment by idiot-z** [comment]
Hi @Speierers 
Thank you for your prompt response. Just to be sure, is differentiation of weights for blended bsdf supposed to be functional in the current release (fe2a1bd)? Because there is a star next to the corresponding parameter, however I get very strange results.

**Comment by Speierers** [comment]
Hi @idiot-z ,
Please open a seperate issue for this as it is completely unrelated to this PR ;)

**Comment by mehrab2603** [comment]
Is there any plan of merging this with master? I need some features from master (projector emitter) that were implemented after this. Tried merging myself but got so many conflicts that I don't know how to fix.

**Comment by Speierers** [comment]
@mehrab2603 No we are not going to merge this with `master` as we are about to release our refactored/redesigned version of the codebase which will render this PR completely out of sync.

In the future, we are planning on implementing integrators that can deal with discontinuities. But this won't be part of the upcoming release unfortunately.

**Review by Scutua — APPROVED** [review]
